### DISCUSSION TITLE
Stepper

### DISCUSSION THREAD

**Discussion by Febakke** [digdir/designsystemet#1820] [Components]
## Related issues or relevant information?
- https://github.com/Altinn/app-frontend-react/issues/97
- https://github.com/Altinn/app-frontend-react/issues/188 (Progress indicator made for "Elevundersøkelsen")

### Related research: 
- https://www.linkedin.com/posts/vitalyfriedman_ux-design-activity-7127577055029796864-0JVK?utm_source=share&utm_medium=member_ios
### Other Stepper components
- Carbon progress-indicator: https://carbondesignsystem.com/components/progress-indicator/usage/
- Carbon progress-bar: https://carbondesignsystem.com/components/progress-bar/usage/
- Gov UK: https://design-system.service.gov.uk/patterns/step-by-step-navigation/
- Fælles designsystem: https://designsystem.dk/komponenter/trinindikator/
- Entur: https://design.entur.org/komponenter/navigasjon/stepper
- Nav: https://aksel.nav.no/komponenter/core/stepper
- Atlassian: https://atlassian.design/components/progress-tracker/examples (See also progress bar and progress indicator)
- Mattilsynet: https://design.mattilsynet.no/?path=/docs/designsystem-steps--docs&globals=theme:Light+large
- Gjensidige: https://www.gjensidige.builders/docs/core/?path=/story/packages-builders-components-progress--step-horizontal

## Specification and requirements
- _Add specification and requirements_

Issue: https://github.com/digdir/designsystemet/issues/85

**Comment by mrosvik** [comment]
See examples from Enova and Udir in Figma: 
<img width="1077" height="1040" alt="image" src="https://github.com/user-attachments/assets/53c24084-8fda-4e6c-aaa2-adbe8a37ae07" />

**Comment by eirikbacker** [comment]
In Mattilsynet, we experience stepper is not necessarily the correct pattern to show progress of forms, as it takes a lot of space, needs to vary between mobile and desktop and can be confusing if steps appear/disapear. Stepper is very practical to show timelines though.
Inspiration from Fortum:

<img width="492" height="491" alt="image" src="https://github.com/user-attachments/assets/2e828e1e-5e0f-4445-acd0-fb50177c2d86" />

**Comment by Camulos** [comment]
Våpenregisteret - used in for example forms to edit weapons or weapon-parts 

<img width="422" height="168" alt="Skjermbilde 2026-02-25 kl  13 42 50" src="https://github.com/user-attachments/assets/0897156b-76af-4800-9429-eed97dfabc8b" />

<img width="411" height="685" alt="Skjermbilde 2026-02-25 kl  13 45 32" src="https://github.com/user-attachments/assets/7839080e-f222-4769-a308-11db28b4f349" />

<img width="947" height="241" alt="Skjermbilde 2026-02-25 kl  13 45 37" src="https://github.com/user-attachments/assets/09e89ea2-1c8a-4dd9-bde1-9fdc07b6d90c" />

**Comment by Camulos** [comment]
Concept: Subscription service for kunngjøringer, used in form to set up new subscription.
Desktop
<img width="384" height="440" alt="Skjermbilde 2026-02-25 kl  13 50 08" src="https://github.com/user-attachments/assets/9601c5ee-1543-4a1e-ba94-5faa4cd93193" />

Mobile
<img width="660" height="264" alt="Skjermbilde 2026-02-25 kl  13 51 48" src="https://github.com/user-attachments/assets/76f61ff6-d6ce-4cee-8af4-f4cd70e1932e" />

**Comment by kimandrefladen** [comment]
For øyeblikket ser Fiskeridirektoratets stepper-aktige komponent slik ut:

Før du åpner et substep:
<img width="2040" height="2218" alt="Før du åpner substep" src="https://github.com/user-attachments/assets/ba1a0491-eb2e-4cfc-9baa-e84b4d2608e4" />


Inne i et substep:
<img width="2040" height="2472" alt="Inne i et substep" src="https://github.com/user-attachments/assets/970eec60-d0c6-4bb0-a554-5504c327befb" />

**Comment by unekinn** [comment]
Vi har eit par variantar i Udir:
- [ProgressBar](https://design.udir.no/beta/?path=/docs/components-progressbar--docs) er kun i bruk i skjema med veldig mange sider med spørsmål, der alle spørsmål er frivillige
- [FormNavigation](https://design.udir.no/beta/?path=/docs/components-formnavigation--docs) er tenkt for dei fleste fleirside-skjema med validering

Vi ser at vi kanskje treng ein slags "tvungen flyt" / wizard som ikkje nødvendigvis er kun skjema, men har ikkje komme så langt som å undersøke det nærare enno.

### ProgressBar

<img width="611" height="600" alt="image" src="https://github.com/user-attachments/assets/dbca77fb-b851-4801-a599-bd5b3bc2828c" />

### FormNavigation

<img width="611" height="795" alt="image" src="https://github.com/user-attachments/assets/5516d1e1-aa29-46e2-bfa6-3b5b566e540a" />

<img width="611" height="475" alt="image" src="https://github.com/user-attachments/assets/13390296-ecf8-45e0-bf37-003f1d2965f2" />

<img width="611" height="475" alt="image" src="https://github.com/user-attachments/assets/990e485f-5f60-4af8-addc-a0f31b5a0b37" />

**Comment by mrosvik** [comment]
Consider whether we should have a page similar to the Datepicker page, where we describe alternative methods for addressing the different needs.

**Comment by Hedda-ro** [comment]
@Camulos recommended this thread - I would like to add that the new solution (in development), "Registrering i Enhetsregisteret og Foretaksregisteret", would benefit from such a stepper component. There is currently a step bar in the solution (section based), but it does not meet the needs arising from the scale and volume of information (objects) involved in current and near-future development.

**Reply by eirikbacker** [reply]
Super to know @Hedda-ro! Do you have any thoughts on vertical vs. horizontal version, requirements (label text? description? clickability? can each step have a status? collapse/expand? Sub processes? etc.) ☺️

**Reply by Hedda-ro** [reply]
Thank you for your feedback and questions!

For this multi-step schema solution, a vertical stepper would perhaps be preferable, considering both scalability (including unknown future needs) and responsive requirements.

Sub-processes also represent an interesting feature to explore, as they could potentially support future concepts such as merging and demerging in relation to limited companies. Another example of a potential use case could be the management of sub-entities, particularly for companies with a large number of sub-entities.

Regarding the other features listed, it is tempting to say, 'Yes, please, all of them'. However, a simpler version would be sufficient for an initial trial in regards to this solution😊

**Most important**
- Label text
- Clickability, including the option to adjust what is clickable depending on the applicable rules (e.g. Altinn task processes including signing)
- Fundamental statuses: error and perhaps visited
- (sub-processes most likely placed here. Concepts related to limited companies, a major planned future delivery, are yet to be fully defined)

**Less important**
- Collapse/expand (unless it comes to sub-processes)
- Description
- Multiple status alternatives, e.g. 'Fulfilled' - sometimes difficult to both define and understand (?)
